modyn/selector/internal/selector_strategies/downsampling_strategies/submodular_downsampling_strategy.py

```diff
@@ -1,32 +1,29 @@
+from modyn.config.schema.downsampling_config import SubmodularDownsamplingConfig
 from modyn.selector.internal.selector_strategies.downsampling_strategies import AbstractDownsamplingStrategy
-from modyn.trainer_server.internal.trainer.remote_downsamplers.deepcore_utils.submodular_function import (
-    SUBMODULAR_FUNCTIONS,
-)
 from modyn.utils import DownsamplingMode
 
 
 class SubmodularDownsamplingStrategy(AbstractDownsamplingStrategy):
-    def __init__(self, downsampling_config: dict, modyn_config: dict, pipeline_id: int, maximum_keys_in_memory: int):
+    def __init__(
+        self,
+        downsampling_config: SubmodularDownsamplingConfig,
+        modyn_config: dict,
+        pipeline_id: int,
+        maximum_keys_in_memory: int,
+    ):
         super().__init__(downsampling_config, modyn_config, pipeline_id, maximum_keys_in_memory)
-
+        self.submodular_function = downsampling_config.submodular_function
+        self.submodular_optimizer = downsampling_config.submodular_optimizer
+        self.selection_batch = downsampling_config.selection_batch
+        self.balance = downsampling_config.balance
         self.remote_downsampling_strategy_name = "RemoteSubmodularDownsamplingStrategy"
 
     def _build_downsampling_params(self) -> dict:
         config = super()._build_downsampling_params()
-
-        if self.downsampling_config.get("submodular_function") is None:
-            raise ValueError(
```

Comment:
I think this might have gotten lost somehow, at least I don't find this list imported from deepcore anymore

Reply:
Sorry can you clarify it more? I didn't get it

Reply:
These submodular functions are in deepcore https://github.com/eth-easl/modyn/blob/6c3d63f22e98f41d3bcbd0c0cc2fcd1053abdc4f/modyn/trainer_server/internal/trainer/remote_downsamplers/deepcore_utils/submodular_function.py#L6

Reply:
I do not find any place where you import and check this. I double checked  and see now that you hardcode the literal values in line 81 of downsampling_config.py. I think rather than hardcoding we should load the list of values from this function. Before, we checked explicitly whether they are supported there. Now if we were to add a new function we'd need to add it both and the schema and there, but it would be good to have the schema load it from the file you link as before.

Reply:
I think it is the right practice to hardcode in the config file. The config file should be the single source of truth on what the system supports, as the config is the user interface and it is definitely difficult for a newbie to dig out where in the actual business code what function is supported.

I can actually delete the check on `RemoteDownsampler` side https://github.com/eth-easl/modyn/blob/6c3d63f22e98f41d3bcbd0c0cc2fcd1053abdc4f/modyn/trainer_server/internal/trainer/remote_downsamplers/remote_submodular_downsampling_strategy.py#L57 together with the `SUBMODULAR_FUNCTIONS` list there, because we know that once the config is validated, the value must be one of supported functions.

But I anyway have to have another PR that replaces the `dict` on the `RemoteDownsampler` side to Pydantic model, so I do not make the effort now. After this next PR (to replace dict on `RemoteDownsampler` side), the single occurence of supported function list is in the config.

Reply:
Mhm, okay, since also in this file the validation is happening on a literal level and not an object/module level, I think you're right. If we would check what functions are available in an object sense I would disagree but let's not overcomplicate this and keep it only in the config file. Thank you!